File: src/typescript/src/transport/transaction/message/outgoing-message.ts
import { IPolyBus } from '../../../i-poly-bus';
import { Message } from './message';
import { MessageInfo } from './message-info';

/**
 * Represents an outgoing message in the transport layer.
 * Contains the message object, its type, and serialized body content.
 */
export class OutgoingMessage extends Message {
  private _body: string;
  private _endpoint: string | undefined;
  private _message: any;
  private _messageInfo: MessageInfo;
  private _messageType: Function;

  /**
   * Creates a new OutgoingMessage instance.
   * @param bus The bus instance associated with the message.
   * @param message The message object to be sent.
   * @param endpoint An optional location to explicitly send the message to.
   * @param messageInfo The message info describing metadata about the message.
   */
  constructor(bus: IPolyBus, message: any, endpoint?: string, messageInfo?: MessageInfo) {
    super(bus);
    this._message = message;
    this._messageType = message?.constructor || Object;
    this._body = '';
    this._endpoint = endpoint;
    this._messageInfo = messageInfo ?? bus.messages.getMessageInfo(message?.constructor) ?? new MessageInfo(
      0 as any,
      '',
      '',
      0,
      0,
      0
    );
  }

  /**
   * The message info describing metadata about the message.
   */
  public get messageInfo(): MessageInfo {
    return this._messageInfo;
  }

  public set messageInfo(value: MessageInfo) {
    this._messageInfo = value;
  }

  /**
   * If the transport supports delayed messages, this is the time at which the message should be delivered.
   */
  public deliverAt: Date | undefined;

  /**
   * Gets the type of the message.
   */
  public get messageType(): Function {
    return this._messageType;
  }

  /**
   * Sets the type of the message.
   */
  public set messageType(value: Function) {
    this._messageType = value;
  }

  /**
   * The serialized message body contents.
   */
  public get body(): string {
    return this._body;
  }

  /**
   * Sets the serialized message body contents.
   */
  public set body(value: string) {
    this._body = value;
  }

  /**
   * An optional location to explicitly send the message to.
   */
  public get endpoint(): string | undefined {
    return this._endpoint;
  }

  /**
   * Sets the endpoint for the message.
   */
  public set endpoint(value: string | undefined) {
    this._endpoint = value;
  }

  /**
   * The message object.
   */
  public get message(): any {
    return this._message;
  }

  /**
   * Sets the message object.
   */
  public set message(value: any) {
    this._message = value;
  }
}

File: src/python/src/transport/transaction/message/outgoing_message.py
from typing import Any, Optional, Type
from datetime import datetime
from src.transport.transaction.message.message import Message
from src.transport.transaction.message.message_info import MessageInfo
from src.i_poly_bus import IPolyBus


class OutgoingMessage(Message):
    """
    Represents an outgoing message to the transport.
    """

    def __init__(
        self,
        bus: "IPolyBus",
        message: Any,
        endpoint: Optional[str] = None,
        message_info: Optional[MessageInfo] = None
    ):
        super().__init__(bus)
        self._message = message
        self._message_type = type(message)
        self._message_info = message_info if message_info is not None else bus.messages.get_message_info(type(message))
        self._body = ""
        self._endpoint = endpoint
        self._deliver_at: Optional[datetime] = None

    @property
    def deliver_at(self) -> Optional[datetime]:
        """
        If the transport supports delayed messages, this is the time at which the message should be delivered.
        """
        return self._deliver_at

    @deliver_at.setter
    def deliver_at(self, value: Optional[datetime]) -> None:
        """
        Set the delivery time for delayed messages.
        """
        self._deliver_at = value

    @property
    def message_info(self) -> Optional[MessageInfo]:
        """
        The message info describing metadata about the message.
        """
        return self._message_info

    @message_info.setter
    def message_info(self, value: Optional[MessageInfo]) -> None:
        """
        Set the message info.
        """
        self._message_info = value

    @property
    def message_type(self) -> Type:
        """
        The type of the message.
        """
        return self._message_type

    @message_type.setter
    def message_type(self, value: Type) -> None:
        """
        Set the message type.
        """
        self._message_type = value

    @property
    def endpoint(self) -> Optional[str]:
        """
        An optional location to explicitly send the message to.
        """
        return self._endpoint

    @endpoint.setter
    def endpoint(self, value: Optional[str]) -> None:
        """
        Set the endpoint.
        """
        self._endpoint = value
    @property
    def endpoint(self) -> str:
        """
        If the message is a command then this is the endpoint the message is being sent to.
        If the message is an event then this is the source endpoint the message is being sent from.
        """
        return self._endpoint

    @endpoint.setter
    def endpoint(self, value: str) -> None:
        """
        Set the endpoint.
        """
        self._endpoint = value

    @property
    def message(self) -> Any:
        """
        The message object.
        """
        return self._message

    @message.setter
    def message(self, value: Any) -> None:
        """
        Set the message object.
        """
        self._message = value
